mirror-main.go

```diff
@@ -245,221 +266,422 @@ func doMirror(sURLs mirrorURLs, progressReader *progressBar, accountingReader *a
-				// Print in new line and adjust to top so that we don't print over the ongoing scan bar
-				if !globalQuiet && !globalJSON {
-					console.Eraseline()
-				}
-				errorIf(sURLs.Error.Trace(), "Unable to prepare URLs for mirroring.")
-				break
-			}
-			if sURLs.isEmpty() {
-				break
-			}
-			jsonData, e := json.Marshal(sURLs)
-			if e != nil {
-				session.Delete()
-				fatalIf(probe.NewError(e), "Unable to marshal URLs into JSON.")
-			}
-			fmt.Fprintln(dataFP, string(jsonData))
-			if !globalQuiet && !globalJSON {
-				// Source content is empty if removal is requested,
-				// put targetContent on to scan bar.
+				// Print in new line and adjust to top so that we
+				// don't print over the ongoing progress bar.
 				if sURLs.SourceContent != nil {
-					scanBar(sURLs.SourceContent.URL.String())
-				} else if sURLs.TargetContent != nil && isRemove {
-					scanBar(sURLs.TargetContent.URL.String())
+					errorIf(sURLs.Error.Trace(sURLs.SourceContent.URL.String()),
+						fmt.Sprintf("Failed to copy ‘%s’.", sURLs.SourceContent.URL.String()))
+				} else {
+					// When sURLs.SourceContent is nil, we know that we have an error related to removing
+					errorIf(sURLs.Error.Trace(sURLs.TargetContent.URL.String()),
+						fmt.Sprintf("Failed to remove ‘%s’.", sURLs.TargetContent.URL.String()))
+				}
+
+				// For all non critical errors we can continue for the
+				// remaining files.
+				switch sURLs.Error.ToGoError().(type) {
+				// Handle this specifically for filesystem related errors.
+				case BrokenSymlink, TooManyLevelsSymlink, PathNotFound, PathInsufficientPermission:
+					continue
+				// Handle these specifically for object storage related errors.
+				case BucketNameEmpty, ObjectMissing, ObjectAlreadyExists, ObjectAlreadyExistsAsDirectory, BucketDoesNotExist, BucketInvalid, ObjectOnGlacier:
+					continue
 				}
+
+				// For critical errors we should exit. Session
+				// can be resumed after the user figures out
+				// the  problem. We know that
+				// there are no current copy actions, because
+				// we're not multi threading now.
+
+				// this issue could be separated using separate
+				// error channel instead of using sURLs.Error
+				ms.CloseAndDie()
+			}
+
+			// finished harvesting urls, save queue to session data
+			if err := ms.queue.Save(ms.NewDataWriter()); err != nil {
+				ms.status.fatalIf(probe.NewError(err), "Unable to save queue.")
 			}
-			// Remember totalBytes only for mirror not for removal,
+
 			if sURLs.SourceContent != nil {
-				totalBytes += sURLs.SourceContent.Size
+				ms.Header.LastCopied = sURLs.SourceContent.URL.String()
+			} else if sURLs.TargetContent != nil {
+				ms.Header.LastRemoved = sURLs.TargetContent.URL.String()
 			}
-			totalObjects++
-		case <-trapCh:
-			// Print in new line and adjust to top so that we don't print over the ongoing scan bar
-			if !globalQuiet && !globalJSON {
-				console.Eraseline()
+
+			ms.Save()
+
+			if sURLs.SourceContent != nil {
```

Comment:
the sURLs.SourceContent is being used to determine if the file is being deleted or created. Depending on that it will set LastCopied or LastRemoved.